batch/batch/worker.py

```diff
@@ -81,30 +84,94 @@ def free(self, port):
         except DockerError as e:
-            # 408 request timeout, 503 service unavailable
-            if e.status == 408 or e.status == 503:
-                log.exception(f'in docker call to {f.__name__}, retrying', stack_info=True)
-            # DockerError(500, 'Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
-            # DockerError(500, 'error creating overlay mount to /var/lib/docker/overlay2/545a1337742e0292d9ed197b06fe900146c85ab06e468843cd0461c3f34df50d/merged: device or resource busy'
-            elif e.status == 500 and ("request canceled while waiting for connection" in e.message or
-                                      re.match("error creating overlay mount.*device or resource busy", e.message)):
-                log.exception(f'in docker call to {f.__name__}, retrying', stack_info=True)
-            else:
+            # exponentially back off, up to (expected) max of 30s
+            t = delay * random.random()
+            await asyncio.sleep(t)
+            delay = min(delay * 2, 60.0)
+
+            error += 1
+            if error == 5:
+                log.exception(f'encountered 5 errors while creating container {name}, aborting', stack_info=True)
                 raise
-        except aiohttp.client_exceptions.ServerDisconnectedError:
-            log.exception(f'in docker call to {f.__name__}, retrying', stack_info=True)
-        except asyncio.TimeoutError:
-            log.exception(f'in docker call to {f.__name__}, retrying', stack_info=True)
-        # exponentially back off, up to (expected) max of 30s
-        t = delay * random.random()
-        await asyncio.sleep(t)
-        delay = min(delay * 2, 60.0)
+
+            # 409 container with name already exists
+            if e.status == 409:
+                try:
+                    return await docker.containers.get(name)
+                except DockerError as eget:
+                    # 404 No such container
+                    if eget.status == 404:
+                        continue
+                    raise
+            raise
+
+
+async def start_container(container):
+    try:
+        return await container.start()
+    except DockerError as e:
+        # 304 container has already started
+        if e.status == 304:
+            return
+        raise
+
+
+async def stop_container(container):
+    try:
+        return await container.stop()
+    except DockerError as e:
+        # 304 container has already stopped
+        if e.status == 304:
+            return
+        raise
+
+
+async def delete_container(container, *args, **kwargs):
+    try:
+        return await container.delete(*args, **kwargs)
+    except DockerError as e:
+        # 404 container does not exist
+        # 409 removal of container is already in progress
+        if e.status in (404, 409):
+            return
+        raise
```

Comment:
I verified these seem to catch the present and reasonable-to-retry errors in [the docker API v1.35](https://docs.docker.com/engine/api/v1.35/#operation/ContainerDelete) which is the one used by [aiodocker v0.14.0](https://github.com/aio-libs/aiodocker/blob/v0.14.0/aiodocker/docker.py#L64) and 0.17.0 as well.